core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java

```diff
@@ -271,25 +277,47 @@ public void process(StreamDecoderInput in, StreamDecoderOutput<WebSocketFrame> o
                         continue; // to while loop
                     }
 
+                    if (frameOpcode != WebSocketFrameType.TEXT.opcode() &&
+                        frameOpcode != WebSocketFrameType.BINARY.opcode() &&
+                        frameOpcode != WebSocketFrameType.CONTINUATION.opcode()) {
+                        throw protocolViolation(WebSocketCloseStatus.INVALID_MESSAGE_TYPE,
+                                                "Cannot decode a web socket frame with opcode: " + frameOpcode);
+                    }
+
                     final WebSocketFrame decodedFrame;
                     if (frameOpcode == WebSocketFrameType.TEXT.opcode()) {
                         decodedFrame = WebSocketFrame.ofPooledText(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
                     } else if (frameOpcode == WebSocketFrameType.BINARY.opcode()) {
                         decodedFrame = WebSocketFrame.ofPooledBinary(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
-                    } else if (frameOpcode == WebSocketFrameType.CONTINUATION.opcode()) {
-                        decodedFrame = WebSocketFrame.ofPooledContinuation(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
                     } else {
-                        throw protocolViolation(WebSocketCloseStatus.INVALID_MESSAGE_TYPE,
-                                                "Cannot decode a web socket frame with opcode: " + frameOpcode);
+                        assert frameOpcode == WebSocketFrameType.CONTINUATION.opcode();
+                        decodedFrame = WebSocketFrame.ofPooledContinuation(payloadBuffer, finalFragment);
                     }
                     logger.trace("{} is decoded.", decodedFrame);
+
                     if (finalFragment) {
                         fragmentedFramesCount = 0;
+                        if (aggregatingFrames.isEmpty()) {
+                            out.add(decodedFrame);
+                        } else {
+                            aggregatingFrames.add(decodedFrame);
```

Comment:
From my understanding, `maxFramePayloadLength` does not limit the total length of `aggregatingFrames`. I guess it could be a vulnerable point. 

What do you think of applying `maxFramePayloadLength` to the `aggregatingFrames` if `aggregateContinuation` is enabled? If reasonable, it is also worth noting on the Javadoc of `aggregateContinuation` option.

Reply:
That makes sense and is a good suggestion. 👍